online-shop/src/app/data.service.ts

```diff
@@ -0,0 +1,142 @@
+import { HttpClient, HttpErrorResponse } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { Observable, throwError } from 'rxjs';
+import { IProduct } from './product-list/product-list.component';
+import { catchError } from 'rxjs/operators';
+
+export interface ICartProduct {
+  productId: number,
+  category: string,
+  name: string,
+  price: number,
+  quantity: number,
+}
+
+interface IOrder {
+  customer: string,
+  products: ICartProduct[],
+}
+
+export interface IUser {
+  username: string,
+  password: string | null,
+  fullName?: string,
+  roles?: string[]
+}
+
+
+@Injectable({
+  providedIn: 'root'
+})
+export class DataService {
+  private shoppingCart: ICartProduct[] = [];
+  private BASE_URL = "http://localhost:3000";
+  private currentUser: IUser = { username: "", password: null };
+  private isCustomer: boolean = false;
+  private isAdmin: boolean = false;
+
+
+  constructor(private http: HttpClient) { }
+
+  private handleError(error: HttpErrorResponse) {
+    if (error.status === 0) {
+      // A client-side or network error occurred. Handle it accordingly.
+      console.error('An error occurred:', error.error);
+      return throwError(error.error);
+    } else {
+      // The backend returned an unsuccessful response code.
+      // The response body may contain clues as to what went wrong.
+      console.error(
+        `Backend returned code ${error.status}, body was: `, error.error);
+      return throwError(error.statusText);
+    }
+  }
+
+  addProductToShoppingCart(prod: ICartProduct): void {
+    this.shoppingCart.push(prod);
+  }
+
+  removeItemFromCart(id: number): void {
+    this.shoppingCart = this.shoppingCart.filter(prod => prod.productId != id);
+  }
+
+  sendOrder(): Observable<any> {
```

Comment:
Try to avoid using any and add a specific type. any is used only for special use cases where it's not possible to have a specific type